### PR TITLE
Parametrize ObjectiveFunction

### LINKED ISSUE #197: get ObjectiveFunction is not type stable
We can query the function type...

### PR DESCRIPTION
**Pull request by blegat**
Let's break the MOI world again \o/
Closes #197

### CODE REVIEW

**Comment by mlubin** [comment]
This is a usability loss for MOI. It would be nice if you could still use `ObjectiveFunction()` in a `set!` call, because the type information is already encoded in the object being set. Similarly if you don't care about the type, you might want to use `ObjectiveFunction()` in `get`.

Instead of making a special case for `ObjectiveFunction`, what about trying to address https://github.com/JuliaOpt/MathOptInterface.jl/issues/31?

**Comment by blegat** [comment]
We could have a default constructor of ObjectiveFunction:
```julia
ObjectiveFunction() = ObjectiveFunction{AbstractScalarFunction}()
```
and define this so that solver instance do not need to worry about it
```julia
set!(instance::AbstractInstance, ::ObjectiveFunction{AbstractScalarFunction}, f::F) = set!(instance, ObjectiveFunction{F}, f)
```

**Comment by blegat** [comment]
For all attributes (except `RawSolver` and `ObjectiveFunction`), the return type can be determined from the documentation. For `ListOfConstraints`, it is `Vector{Tuple{AbstractFunction, AbstractSet}}` for instance. This PR allows `ObjectiveFunction` not to be an exception.

**Comment by mlubin** [comment]
> For all attributes (except RawSolver and ObjectiveFunction), the return type can be determined from the documentation.

Not quite, we're ambiguous about the types of numbers, for example. But this is a practical fix for the current issues, so I'll support it.

**Comment by blegat** [comment]
> Not quite, we're ambiguous about the types of numbers, for example. But this is a practical fix for the current issues, so I'll support it.

Yes, in most cases, solver support just one number type as input and output the same type but we want to remain flexible :-P

File: src/attributes.jl [hunk 1]
```diff
@@ -252,12 +252,15 @@ has value greater than zero.
 struct ListOfConstraints <: AbstractInstanceAttribute end
 
 """
-    ObjectiveFunction()
+    ObjectiveFunction{F<:AbstractFunction}()
```
Comment (mlubin) [inline src/attributes.jl]: `AbstractFunction` or `AbstractScalarFunction`?

**Comment by mlubin** [comment]
We should experiment with https://github.com/JuliaOpt/MathOptInterface.jl/pull/198#issuecomment-358971403 before merging this PR.

**Comment by blegat** [comment]
It turns out
```julia
set!(instance::AbstractInstance, ::ObjectiveFunction{AbstractScalarFunction}, f::F) = set!(instance, ObjectiveFunction{F}(), f)
```
creates an ambiguity with `set(::InstanceType, ::AbstractInstanceAttribute, value)` that most intermediate layer (such as InstanceManager) define.

**Comment by blegat** [comment]
I think it is better not to do this trick and force the user to do
```julia
MOI.set!(instance, MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}, ...)
```
It uses a few additional keystrokes but it is clean and MOI will most often be used in package so more keystroke is no issue. For quick scripts and command line, people use JuMP where keystroke matters.

**Comment by mlubin** [comment]
What if we have attributes called `TypedObjectiveFunction{F}()` and `ObjectiveFunction()`? `ObjectiveFunction()` would be a shortcut for `TypedObjectiveFunction` and would not show up in `ListOfInstanceAttributesSet`.

**Comment by blegat** [comment]
`set!(instance, ObjectiveFunction(), ...)` will be `set!(instance, ::Type{TypedObjectiveFunction})` ?
We could also add `setobjective!` as a syntactic sugar for MOI user (implementations of MOI shouldn't care about it
```julia
function setobjective!(instance::AbstractInstance, sense::OptimizationSense, f::AbstractScalarFunction)
    MOI.set!(instance, MOI.ObjectiveSense(), sense)
    MOI.set!(instance, MOI.ObjectiveFunction{typeof(f)}, f)
end
```
Setting these two attributes separately from the user feels weird anyway.

**Comment by mlubin** [comment]
Ok, we can merge the PR as is or with `setobjective!`, up to you. **But** first I want to see PRs queued up on MOIU, MOIT, and JuMP with the corresponding fixes so that the MOI ecosystem will only be in a broken state for a short period of time.

**Comment by blegat** [comment]
All PRs are queued up :) I have chosen not to add `setobjective!` for now. It would have maybe made things a bit simpler in MOIU and JuMP (it can be added later smoothly without breaking anything) but in MOIT, I think we shouldn't use `setobjective!` as it would allow solver to pass tests by only implementing this function and not the `set!` equivalent.

**Pull request merged**